src/Features/LanguageServer/Protocol/Handler/MapCode/MapCodeHandler.cs

```diff
@@ -0,0 +1,137 @@
+    [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+    public MapCodeHandler()
+    {
+    }
+
+    public bool MutatesSolutionState => false;
+    public bool RequiresLSPSolution => true;
+
+    public async Task<WorkspaceEdit?> HandleRequestAsync(MapCodeParams request, RequestContext context, CancellationToken cancellationToken)
+    {
+        //TODO: handle request.Updates if not empty
+        Contract.ThrowIfNull(context.Solution);
+
+        using var _ = PooledDictionary<Uri, TextEdit[]>.GetInstance(out var uriToEditsMap);
+        foreach (var codeMapping in request.Mappings)
+        {
+            var mappingResult = await MapCodeAsync(codeMapping).ConfigureAwait(false);
+
+            // Assume no two codeMappings target a common document.
+            if (mappingResult is (Uri uri, TextEdit[] textEdits))
+                uriToEditsMap[uri] = textEdits;
+        }
+
+        // return a combined WorkspaceEdit
+        if (context.GetRequiredClientCapabilities().Workspace?.WorkspaceEdit?.DocumentChanges is true)
+        {
+            return new WorkspaceEdit
+            {
+                DocumentChanges = uriToEditsMap.Select(kvp => new TextDocumentEdit
+                {
+                    TextDocument = new OptionalVersionedTextDocumentIdentifier { Uri = kvp.Key },
+                    Edits = kvp.Value,
+                }).ToArray()
+            };
+        }
+        else
+        {
+            return new WorkspaceEdit
+            {
+                Changes = uriToEditsMap.ToDictionary(kvp => ProtocolConversions.GetDocumentFilePathFromUri(kvp.Key), kvp => kvp.Value)
+            };
+        }
+
+        async Task<(Uri, TextEdit[])?> MapCodeAsync(LSP.MapCodeMapping codeMapping)
+        {
+            var textDocument = codeMapping.TextDocument;
+            if (textDocument == null)
+                return null;
+
+            if (context.Solution.GetDocument(textDocument) is not Document document)
+                return null;
+
+            var codeMapper = document.GetLanguageService<IMapCodeService>();
+            if (codeMapper == null)
+                return null;
+
+            var focusLocations = await ConvertFocusLocationsToDocumentAndSpansAsync(
+                document,
+                textDocument,
+                codeMapping.FocusLocations,
+                cancellationToken).ConfigureAwait(false);
+
+            var newDocument = await codeMapper.MapCodeAsync(
+                document,
+                codeMapping.Contents.ToImmutableArrayOrEmpty(),
+                focusLocations,
+                formatMappedCode: true,
+                cancellationToken).ConfigureAwait(false);
+
+            if (newDocument is null)
+                return null;
+
+            var textChanges = (await newDocument.GetTextChangesAsync(document, cancellationToken).ConfigureAwait(false)).ToImmutableArray();
+            if (textChanges.IsEmpty)
+                return null;
```

Comment:
If the mapper did not map anything, does it return the unchanged document, or null?  If it is the latter then this should throw as an empty set of changes would be unexpected

Reply:
the way the mapCode request is defined today, it can include multiple map requests, so if we failed to map one of them, we'd still try provide results for others. I will log a warning here